### DISCUSSION TITLE
Get ready to share your plugins on the IF Explorer!

### DISCUSSION THREAD

**Discussion by jmcook1186** [Green-Software-Foundation/if#717] [Announcements]
Hi plugin builders!

🎉 We are aiming to launch our IF Explorer website by the end of the month! 🎉

This site will be used to share plugins so users can browse or search for your plugins to support their IF use-cases.

You can submit information about your plugin now so that they are listed on launch!

## How to submit your plugin

Fill in this form!

[Plugin submission form](https://wiki.greensoftware.foundation/impact-framework-plugins)

You can suggest tags, or leave them blank. We might retroactively apply tags from a predefined list shortly before launch.

Providing the requested data will enable us to render a card presenting your plugin in the UI.

You can start gathering these data and/or submit the form now so your plugin is on the UI ready for launch!

## Acceptance criteria

As long as you provide valid answers to the questions on the form, we don't see any obviously dishonest behaviour and your plugin seems to work then your plugin will get listed - we are only going to do the minimum of QC for now - but we will not hesitate to delist plugins if they turn out to be broken or somehow malicious. 

We also intend to create some badges to highlight projects that have gone the extra mile in testing, documenting and/or providing citations to support their approach - more on that later.

It is really important to have detailed documentation in your README, an example manifest file that runs your plugin successfully, and unit tests.

Your plugin documentation should include the following information:

- A list of the required and optional parameters and return values, with types and any valid ranges/properties clearly defined.
- A list of required and optional configuration
- A description of any environment setup such as credentials in environment variables, etc.
- Installation instructions
- A written description of the plugin behaviour
- A link to your unit tests 
- A list of errors that your plugin can raise, the behaviours that cause them and potential remedies.
- A demo manifest that executes the plugin correctly without errors
- A reference list of any publications or other material supporting the approach you have taken in your plugin.

## Plugin repo structure 

Please submit the form once for each individual plugin - i.e. **do not** submit one form for a repo containing multiple plugins - split the plugins across multiple repositories.

We would much prefer each plugin to have its own repository and independent documentation. This means we can provide one card per plugin on the website and users will be able to find the plugin they are looking for more easily.

## Hackathon plugins

There were lots of exciting plugins created during the recent hackathon. 
Pinging some hackathon plugin builders below!!

@mgriffin-scottlogic
@szm100
@teto1992
@hoernschen
@dukeofjukes
@mgriffin-scottlogic
@jcendava
@moin-oss
@tianbiao
@kiran-G1
@adamaucamp
@andreic94
@ramgsuri
@srini1978
@ktg9
@pazbardanl
@greenscale-nandesh
@jawache

**Comment by mgriffin-scottlogic** [comment]
Exciting stuff!

Just wanted to confirm, is one plugin per repo preferred or mandatory?

I can understand that the form is set up to deal with only one plugin but I wasn't sure if 4 differently named plugins from the same repo would be accepted via multiple form submissions.

**Reply by jmcook1186** [reply]
Hi - I think in general it is better to have one repository per plugin, but really the important thing is that there is one *card on the website* per plugin. So as long as you submit the form once for each individual plugin, I think its probably ok.

**Reply by mgriffin-scottlogic** [reply]
Good to know, I think it would be better in the long run but I might not get the chance to split them up before the end of the month.

**Reply by jmcook1186** [reply]
yea that's fair enough!

**Comment by andrew-woosnam** [comment]
Hey @jmcook1186 (+others), do you have any plans to make this site queryable via API? Asking because @josh-swerdlow and I are working on [this VS Code extension](https://github.com/andrew-woosnam/impact-framework-vscode/tree/main) for IF and it would be awesome if we had a robust way to list available IF plugins in a sidebar.

**Reply by jawache** [reply]
Wow amazing work @andrew-woosnam!

@jmcook1186 and @osamajandali I believe we're driving the website from a Notion DB. @osamajandali I'm "assuming" that when the site gets built it will generate a JSON file in the repo with all the relevant plugin data right? Could @andrew-woosnam just query that json export?

**Reply by andrew-woosnam** [reply]
That might be good enough for a start! Honestly anything would be an improvement over the current (i.e. hardcoded) solution 

Btw @jawache we're wondering if this VSC extension is worth pursuing further, so any [feedback](https://github.com/Green-Software-Foundation/if/discussions/737#discussion-6719004) would be very welcome :)

**Comment by jmcook1186** [comment]
Hi folks, we've got a new Ui for the plugin submission - you can access it here:

https://wiki.greensoftware.foundation/how-to-add-plugins

The website will be going live soon, so get your plugins submitted asap if you want them to be included on launch day!

**Comment by andrew-woosnam** [comment]
Hey @jmcook1186 @jawache I came across a potential security vulnerability on the plugins site while working on this [vscode extension](https://github.com/andrew-woosnam/impact-framework-vscode/tree/main). May be unconcerning in the end but I'd like to be sure. Is there a better (i.e. less public) place to discuss futher?

**Reply by jmcook1186** [reply]
Following up by email using the address shared on your GH profile.

**Comment by moin-oss** [comment]
Hi @jmcook1186 , I submitted https://github.com/moin-oss/datadog-importer to the IF Explorer last week and haven't seen this posted on the explorer or received any sort of update.  What's the turnaround for having a submission posted, or receiving feedback on it?

**Reply by jmcook1186** [reply]
oh - thanks @moin-oss - i missed it - i'll get it listed on the site asap! Aiming for Monday.

**Reply by jmcook1186** [reply]
hi @moin-oss - sorry I missed it before - I checked out the submission - it should be live on the explorer site in the next couple of hours.

**Reply by moin-oss** [reply]
Hey @jmcook1186 , no worries, thanks!  I'll keep an eye for it.

**Reply by moin-oss** [reply]
Hey @jmcook1186 , just following up on this.  I haven't seen this show up in the explorer yet; bringing this to your attention in case there's something you missed.

**Reply by jmcook1186** [reply]
oh - i approved it the other day and it's supposed to happen automatically on approval - i'll force a rebuild and hopefully that fixes it.

**Reply by jmcook1186** [reply]
It's there now :o)
https://explorer.if.greensoftware.foundation/

**Reply by moin-oss** [reply]
Thanks, I can see it now!  For future reference, in order to update this entry, would I have to re-submit it, or is there a separate process for that?

**Reply by jmcook1186** [reply]
We don't have a specific process for that at the moment, we're trying to encourage open source moderation of the plugins and expect that if your plugin breaks or does something other than what's described in the docs, then there will be some back and forth between you and your users on your repo, and if things escalate we'll hear about it and delist if necessary. If there are tweaks you'd like to make to the wording, title, links etc, then you can just ping us directly.
